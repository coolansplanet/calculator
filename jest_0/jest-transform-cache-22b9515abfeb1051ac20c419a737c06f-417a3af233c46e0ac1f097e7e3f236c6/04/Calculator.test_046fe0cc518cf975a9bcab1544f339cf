d6aaf0814e14a9793134fe46473d785f
"use strict";

var _interopRequireDefault = require("/home/hermanschmidt/Desktop/calculator/node_modules/@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("/home/hermanschmidt/Desktop/calculator/node_modules/@babel/runtime/helpers/interopRequireWildcard");

var _Calculator = _interopRequireDefault(require("./Calculator"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _enzyme = _interopRequireWildcard(require("enzyme"));

var _enzymeAdapterReact = _interopRequireDefault(require("enzyme-adapter-react-16"));

var _jsxFileName = "/home/hermanschmidt/Desktop/calculator/src/components/Calculator/Calculator.test.js";

_enzyme.default.configure({
  adapter: new _enzymeAdapterReact.default()
});

describe('Calculator component', () => {
  it('The first displayed number should be 0', () => {
    const wrapper = (0, _enzyme.shallow)(_react.default.createElement(_Calculator.default, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 13
      },
      __self: void 0
    }));
    const display = wrapper.find('.display').text();
    expect(display).toEqual('0');
  });
  it('Checking all numbers buttons work', () => {
    const wrapper = (0, _enzyme.shallow)(_react.default.createElement(_Calculator.default, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 19
      },
      __self: void 0
    }));
    const display = wrapper.find('.display').text();
    wrapper.find('.b7').simulate('click');
    setTimeout(() => {
      expect(display).toEqual('7');
    }, 1000);
  });
});
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIkNhbGN1bGF0b3IudGVzdC5qcyJdLCJuYW1lcyI6WyJFbnp5bWUiLCJjb25maWd1cmUiLCJhZGFwdGVyIiwiQWRhcHRlciIsImRlc2NyaWJlIiwiaXQiLCJ3cmFwcGVyIiwiZGlzcGxheSIsImZpbmQiLCJ0ZXh0IiwiZXhwZWN0IiwidG9FcXVhbCIsInNpbXVsYXRlIiwic2V0VGltZW91dCJdLCJtYXBwaW5ncyI6Ijs7Ozs7O0FBQUE7O0FBQ0E7O0FBQ0E7O0FBQ0E7O0FBQ0E7Ozs7QUFHQUEsZ0JBQU9DLFNBQVAsQ0FBaUI7QUFBRUMsRUFBQUEsT0FBTyxFQUFFLElBQUlDLDJCQUFKO0FBQVgsQ0FBakI7O0FBRUFDLFFBQVEsQ0FBQyxzQkFBRCxFQUF5QixNQUFNO0FBRW5DQyxFQUFBQSxFQUFFLENBQUMsd0NBQUQsRUFBMkMsTUFBTTtBQUMvQyxVQUFNQyxPQUFPLEdBQUcscUJBQVEsNkJBQUMsbUJBQUQ7QUFBQTtBQUFBO0FBQUE7QUFBQTtBQUFBO0FBQUEsTUFBUixDQUFoQjtBQUNBLFVBQU1DLE9BQU8sR0FBR0QsT0FBTyxDQUFDRSxJQUFSLENBQWEsVUFBYixFQUF5QkMsSUFBekIsRUFBaEI7QUFDQUMsSUFBQUEsTUFBTSxDQUFDSCxPQUFELENBQU4sQ0FBZ0JJLE9BQWhCLENBQXdCLEdBQXhCO0FBQ0gsR0FKQyxDQUFGO0FBTUFOLEVBQUFBLEVBQUUsQ0FBQyxtQ0FBRCxFQUFzQyxNQUFNO0FBQzFDLFVBQU1DLE9BQU8sR0FBRyxxQkFBUSw2QkFBQyxtQkFBRDtBQUFBO0FBQUE7QUFBQTtBQUFBO0FBQUE7QUFBQSxNQUFSLENBQWhCO0FBQ0EsVUFBTUMsT0FBTyxHQUFHRCxPQUFPLENBQUNFLElBQVIsQ0FBYSxVQUFiLEVBQXlCQyxJQUF6QixFQUFoQjtBQUVBSCxJQUFBQSxPQUFPLENBQUNFLElBQVIsQ0FBYSxLQUFiLEVBQW9CSSxRQUFwQixDQUE2QixPQUE3QjtBQUNBQyxJQUFBQSxVQUFVLENBQUMsTUFBSztBQUNaSCxNQUFBQSxNQUFNLENBQUNILE9BQUQsQ0FBTixDQUFnQkksT0FBaEIsQ0FBd0IsR0FBeEI7QUFFSCxLQUhTLEVBR1AsSUFITyxDQUFWO0FBS0gsR0FWQyxDQUFGO0FBV0gsQ0FuQk8sQ0FBUiIsInNvdXJjZXNDb250ZW50IjpbImltcG9ydCBDYWxjdWxhdG9yIGZyb20gJy4vQ2FsY3VsYXRvcic7XG5pbXBvcnQgUmVhY3QgZnJvbSAncmVhY3QnO1xuaW1wb3J0IFJlYWN0RE9NIGZyb20gJ3JlYWN0LWRvbSc7XG5pbXBvcnQgRW56eW1lIGZyb20gJ2VuenltZSc7XG5pbXBvcnQgQWRhcHRlciBmcm9tICdlbnp5bWUtYWRhcHRlci1yZWFjdC0xNic7XG5pbXBvcnQgeyBzaGFsbG93IH0gZnJvbSAnZW56eW1lJztcblxuRW56eW1lLmNvbmZpZ3VyZSh7IGFkYXB0ZXI6IG5ldyBBZGFwdGVyKCkgfSk7XG5cbmRlc2NyaWJlKCdDYWxjdWxhdG9yIGNvbXBvbmVudCcsICgpID0+IHtcblxuICAgIGl0KCdUaGUgZmlyc3QgZGlzcGxheWVkIG51bWJlciBzaG91bGQgYmUgMCcsICgpID0+IHtcbiAgICAgICAgY29uc3Qgd3JhcHBlciA9IHNoYWxsb3coPENhbGN1bGF0b3IvPik7XG4gICAgICAgIGNvbnN0IGRpc3BsYXkgPSB3cmFwcGVyLmZpbmQoJy5kaXNwbGF5JykudGV4dCgpO1xuICAgICAgICBleHBlY3QoZGlzcGxheSkudG9FcXVhbCgnMCcpO1xuICAgIH0pO1xuXG4gICAgaXQoJ0NoZWNraW5nIGFsbCBudW1iZXJzIGJ1dHRvbnMgd29yaycsICgpID0+IHtcbiAgICAgICAgY29uc3Qgd3JhcHBlciA9IHNoYWxsb3coPENhbGN1bGF0b3IvPik7XG4gICAgICAgIGNvbnN0IGRpc3BsYXkgPSB3cmFwcGVyLmZpbmQoJy5kaXNwbGF5JykudGV4dCgpO1xuXG4gICAgICAgIHdyYXBwZXIuZmluZCgnLmI3Jykuc2ltdWxhdGUoJ2NsaWNrJyk7XG4gICAgICAgIHNldFRpbWVvdXQoKCk9PiB7XG4gICAgICAgICAgICBleHBlY3QoZGlzcGxheSkudG9FcXVhbCgnNycpO1xuXG4gICAgICAgIH0sIDEwMDApO1xuXG4gICAgfSk7XG59KTtcbiJdfQ==