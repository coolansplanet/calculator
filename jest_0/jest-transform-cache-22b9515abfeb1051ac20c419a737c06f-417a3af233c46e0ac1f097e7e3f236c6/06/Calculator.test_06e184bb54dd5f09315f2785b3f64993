945226116803622fcedf2fe0f7b7e1ea
"use strict";

var _interopRequireDefault = require("/home/hermanschmidt/Desktop/calculator/node_modules/@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("/home/hermanschmidt/Desktop/calculator/node_modules/@babel/runtime/helpers/interopRequireWildcard");

var _Calculator = _interopRequireDefault(require("./Calculator"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _enzyme = _interopRequireWildcard(require("enzyme"));

var _enzymeAdapterReact = _interopRequireDefault(require("enzyme-adapter-react-16"));

var _jsxFileName = "/home/hermanschmidt/Desktop/calculator/src/components/Calculator/Calculator.test.js";

_enzyme.default.configure({
  adapter: new _enzymeAdapterReact.default()
});

describe('Calculator component', () => {
  it('The first displayed number should be 0', () => {
    const wrapper = (0, _enzyme.shallow)(_react.default.createElement(_Calculator.default, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 13
      },
      __self: void 0
    }));
    const display = wrapper.find('.display').text();
    expect(display).toEqual('0');
  });
  it('The first displayed number should be 0', () => {
    const wrapper = (0, _enzyme.shallow)(_react.default.createElement(_Calculator.default, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 19
      },
      __self: void 0
    }));
    wrapper.find('.b-7').simulate('click');
    expect(display).toEqual('7');
  });
});
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIkNhbGN1bGF0b3IudGVzdC5qcyJdLCJuYW1lcyI6WyJFbnp5bWUiLCJjb25maWd1cmUiLCJhZGFwdGVyIiwiQWRhcHRlciIsImRlc2NyaWJlIiwiaXQiLCJ3cmFwcGVyIiwiZGlzcGxheSIsImZpbmQiLCJ0ZXh0IiwiZXhwZWN0IiwidG9FcXVhbCIsInNpbXVsYXRlIl0sIm1hcHBpbmdzIjoiOzs7Ozs7QUFBQTs7QUFDQTs7QUFDQTs7QUFDQTs7QUFDQTs7OztBQUdBQSxnQkFBT0MsU0FBUCxDQUFpQjtBQUFFQyxFQUFBQSxPQUFPLEVBQUUsSUFBSUMsMkJBQUo7QUFBWCxDQUFqQjs7QUFFQUMsUUFBUSxDQUFDLHNCQUFELEVBQXlCLE1BQU07QUFFbkNDLEVBQUFBLEVBQUUsQ0FBQyx3Q0FBRCxFQUEyQyxNQUFNO0FBQy9DLFVBQU1DLE9BQU8sR0FBRyxxQkFBUSw2QkFBQyxtQkFBRDtBQUFBO0FBQUE7QUFBQTtBQUFBO0FBQUE7QUFBQSxNQUFSLENBQWhCO0FBQ0EsVUFBTUMsT0FBTyxHQUFHRCxPQUFPLENBQUNFLElBQVIsQ0FBYSxVQUFiLEVBQXlCQyxJQUF6QixFQUFoQjtBQUNBQyxJQUFBQSxNQUFNLENBQUNILE9BQUQsQ0FBTixDQUFnQkksT0FBaEIsQ0FBd0IsR0FBeEI7QUFDSCxHQUpDLENBQUY7QUFNQU4sRUFBQUEsRUFBRSxDQUFDLHdDQUFELEVBQTJDLE1BQU07QUFDL0MsVUFBTUMsT0FBTyxHQUFHLHFCQUFRLDZCQUFDLG1CQUFEO0FBQUE7QUFBQTtBQUFBO0FBQUE7QUFBQTtBQUFBLE1BQVIsQ0FBaEI7QUFDQUEsSUFBQUEsT0FBTyxDQUFDRSxJQUFSLENBQWEsTUFBYixFQUFxQkksUUFBckIsQ0FBOEIsT0FBOUI7QUFDQUYsSUFBQUEsTUFBTSxDQUFDSCxPQUFELENBQU4sQ0FBZ0JJLE9BQWhCLENBQXdCLEdBQXhCO0FBQ0gsR0FKQyxDQUFGO0FBS0gsQ0FiTyxDQUFSIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0IENhbGN1bGF0b3IgZnJvbSAnLi9DYWxjdWxhdG9yJztcbmltcG9ydCBSZWFjdCBmcm9tICdyZWFjdCc7XG5pbXBvcnQgUmVhY3RET00gZnJvbSAncmVhY3QtZG9tJztcbmltcG9ydCBFbnp5bWUgZnJvbSAnZW56eW1lJztcbmltcG9ydCBBZGFwdGVyIGZyb20gJ2VuenltZS1hZGFwdGVyLXJlYWN0LTE2JztcbmltcG9ydCB7IHNoYWxsb3cgfSBmcm9tICdlbnp5bWUnO1xuXG5Fbnp5bWUuY29uZmlndXJlKHsgYWRhcHRlcjogbmV3IEFkYXB0ZXIoKSB9KTtcblxuZGVzY3JpYmUoJ0NhbGN1bGF0b3IgY29tcG9uZW50JywgKCkgPT4ge1xuXG4gICAgaXQoJ1RoZSBmaXJzdCBkaXNwbGF5ZWQgbnVtYmVyIHNob3VsZCBiZSAwJywgKCkgPT4ge1xuICAgICAgICBjb25zdCB3cmFwcGVyID0gc2hhbGxvdyg8Q2FsY3VsYXRvci8+KTtcbiAgICAgICAgY29uc3QgZGlzcGxheSA9IHdyYXBwZXIuZmluZCgnLmRpc3BsYXknKS50ZXh0KCk7XG4gICAgICAgIGV4cGVjdChkaXNwbGF5KS50b0VxdWFsKCcwJyk7XG4gICAgfSk7XG5cbiAgICBpdCgnVGhlIGZpcnN0IGRpc3BsYXllZCBudW1iZXIgc2hvdWxkIGJlIDAnLCAoKSA9PiB7XG4gICAgICAgIGNvbnN0IHdyYXBwZXIgPSBzaGFsbG93KDxDYWxjdWxhdG9yLz4pO1xuICAgICAgICB3cmFwcGVyLmZpbmQoJy5iLTcnKS5zaW11bGF0ZSgnY2xpY2snKTtcbiAgICAgICAgZXhwZWN0KGRpc3BsYXkpLnRvRXF1YWwoJzcnKTtcbiAgICB9KTtcbn0pO1xuIl19