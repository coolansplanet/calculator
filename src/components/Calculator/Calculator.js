import React, { Component } from 'react';
import math from 'mathjs';

class Calculator extends Component {

    constructor(){
        super();
        this.state = {
            display: '0',          //Number shown on the screen
            total: '0',            //Result of previous operations
            operator: null,        //Current operator ( + - * / )
            displayingTotal: true  //Self explanatory flag
        }
    }

    inputNumber( value ){
        /*
            If the previous number shown in the display screen
            was a result (a total), the new input number replaces it.
            Otherwise, the new input number is added
            as a new digit to it.
        */
        let stringeredNumber = value.toString();
        let display =
            this.state.displayingTotal || this.state.display === '0' ?
            stringeredNumber : this.state.display + stringeredNumber;

        /*
            If the previous shown number, the new input number
            and the stored result equals to 0, then the component
            assumes the display is showing a result;
        */
        let displayingTotal =
            this.state.total === '0' &&
            this.state.display === '0' &&
            stringeredNumber === '0';

        /*
            The new input number is added as a new digit
            only if there's available space on the display.
        */
        display =
            display.split('.').join().length <= 11 ?
            display : this.state.display;

        this.setState({
            display: display,
            displayingTotal: displayingTotal
        });
    }

    inputPoint(){
        /*
            Inputs a point only once.
            Also, if the previous number on the display was a result (total),
            clicking the point deletes it, and replaces it by a "0."
        */
        let display;

        if (this.state.displayingTotal) {
            display = '0.';
        }else{
            display = this.state.display;
            display += display.indexOf('.') < 0 ? '.' : '';
        }

        this.setState({
            display: display,
            displayingTotal: false
        });
    }

    inputOperator(inputOperator){

        let nextDisplay;
        let { display, total, operator, displayingTotal } = this.state;

        //Gets the results of the previous operation:
        let previousResult = operator ?
            math.eval(total + operator + display) : total;

        //Gets the new total:
        let nextTotal = operator && !displayingTotal ?
            previousResult.toString() : display;

        //Gets the next number (or value) to show on the display:
        nextDisplay = isNaN(previousResult) ? 'Error' : nextTotal.substr(0, 11);

        //Gets the next operator symbol ( + - * / ):
        let nextOperator = inputOperator === '=' ? null : inputOperator;

        //Updates the state:
        this.setState({
            display: nextDisplay,
            total: nextTotal,
            operator: nextOperator,
            displayingTotal: true
        });
    }


    clear(value){

        //Deletes all the values:
        if (value === 'all') {
            this.setState({
                display: '0',
                total: '0',
                operator: null,
                displayingTotal: true
            });

        //Deletes only the value on the display:
        }else if (value === 'entry'){
            this.setState({
                display: '0'
            });

        //Deletes only one character on the display:
        }else if (value === 'character' && this.state.display !== '0'){
            let display = this.state.display;
            let newDisplay = display.substr(0, display.length - 1);
            newDisplay = display.length > 1 && !isNaN(newDisplay) ? newDisplay : '0';

            this.setState({
                display: newDisplay
            });
        }
    }


    showAbout(){
        //The button "A", when clicked shows the "about" dialog:
        alert('\nCalculator:\nMade with ♥ by Herman Schmidt\n');
    }


    render() {

        const {
            inputNumber,
            inputPoint,
            inputOperator,
            clear
        } = this;

        return (
          <div className="calculator">
            <div className="display">
                {this.state.display}
            </div>

            <div className="buttons">
                <button
                    onClick={inputNumber.bind(this, 7)}
                    className="b7">
                    7
                </button>

                <button
                    onClick={inputNumber.bind(this, 8)}
                    className="b8">
                    8
                </button>

                <button
                    onClick={inputNumber.bind(this, 9)}
                    className="b9">
                    9
                </button>

                <button
                    onClick={inputOperator.bind(this, '/')}
                    className="division">
                    ÷
                </button>

                <button
                    className="special-button clear-all"
                    onClick={clear.bind(this, 'all')}>
                    C
                </button>

                <button
                    onClick={inputNumber.bind(this, 4)}
                    className="b4">
                    4
                </button>

                <button
                    onClick={inputNumber.bind(this, 5)}
                    className="b5">
                    5
                </button>

                <button
                    onClick={inputNumber.bind(this, 6)}
                    className="b6">
                    6
                </button>

                <button
                    onClick={inputOperator.bind(this, '*')}
                    className="multiplication">
                    ×
                </button>

                <button
                    className="special-button clear-entry"
                    onClick={clear.bind(this, 'entry')}>
                    CE
                </button>

                <button
                    onClick={inputNumber.bind(this, 1)}
                    className="b1">
                    1
                </button>

                <button
                    onClick={inputNumber.bind(this, 2)}
                    className="b2">
                    2
                </button>

                <button
                    onClick={inputNumber.bind(this, 3)}
                    className="b3">
                    3
                </button>

                <button
                    onClick={inputOperator.bind(this, '-')}
                    className="substraction">
                    -
                </button>

                <button
                    className="special-button delete"
                    onClick={clear.bind(this, 'character')}>
                    DEL
                </button>

                <button
                    onClick={inputNumber.bind(this, 0)}
                    className="b0">
                    0
                </button>

                <button
                    onClick={inputPoint.bind(this)}
                    className="point">
                    .
                </button>

                <button
                    onClick={inputOperator.bind(this, '=')}
                    className="equals">
                    =
                </button>

                <button
                    onClick={inputOperator.bind(this, '+')}
                    className="addition">
                    +
                </button>

                <button
                    className="special-button about"
                    onClick={this.showAbout}>
                    A
                </button>
            </div>
          </div>
        );
    }
}

export default Calculator;
