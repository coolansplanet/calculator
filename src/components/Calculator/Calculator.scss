.calculator{
    $buttons-size: 54px;
    background: linear-gradient(to bottom right, $orange 50%, $orange-dark, $orange-darker);
    display: inline-block;
    position: relative;
    margin-top: 50px;
    left: 50%;
    transform: translateX(-50%);
    border-radius: 4%;
    padding: $buttons-size/2;
    box-shadow: $buttons-size/4 $buttons-size/4 $buttons-size/2 rgba(black, 0.5);
    z-index: 1;
    outline: none;

    &::after{
        content: '';
        background: linear-gradient(to bottom right, $orange 75%, $orange-dark);
        display: inline-block;
        width: 96%;
        height: 96%;
        position: absolute;
        top: 2%;
        left: 2%;
        border-radius: 4%;
        z-index: -10;

    }

    .display{
        font-family: 'Segment7Standard', monospace;
        font-size: 46px;
        text-align: right;
        background-color: #8d7569;
        color: rgba(black, 0.7);
        height: $buttons-size + 10px;
        line-height: $buttons-size + 10px;
        padding: 0 10px 0 10px;
        display: inline-block;
        width: 100%;
        margin-bottom: $buttons-size/2;
        border-radius: 6px;
        box-shadow: 0 0 10px rgba(black, 0.5) inset;
        text-shadow: 0 0 10px rgba(black, 0.2);
        user-select: none;
    }

    .buttons{
        display: grid;
        grid-template-columns: repeat(5, 1fr);
        grid-gap: $buttons-size/3;
        font-family: 'Roboto', sans-serif;
        font-size: 46px;
        user-select: none;

        button{
            background-color: $orange;
            display: inline-block;
            line-height: $buttons-size;
            height: $buttons-size;
            width: $buttons-size;
            color: white;
            border-radius: 8%;
            position: relative;
            transition: box-shadow 50ms, background-color 50ms;


            &::before{
                //As there's not gradient colors for borders, I had to simulate them with this rule
                content: '';
                background: linear-gradient(to bottom right, $orange-light, $orange, $orange-dark);
                display: inline-block;
                width: 110%;
                height: 110%;
                z-index: -1;
                position: absolute;
                left: -5%;
                top: -5%;
                border-radius: 12%;
                border: 1px solid rgba($orange-darker, 0.5);
                box-shadow: 10px 10px $buttons-size/2 $orange-dark;
            }

            &:hover{

                &::before{
                    border: 1px solid white;
                    box-shadow: 0 0 $buttons-size/2 $orange-light;
                    background: $orange-light;
                }
            }

            &:active{
                box-shadow: none;
                background-color: rgba($orange, 0.5);
                font-size: 44px;

                &::before{
                    border: 1px solid rgba($orange-darker, 0.5);
                    background: linear-gradient(to bottom right, $orange-darker, $orange);
                    box-shadow: none;
                }
            }


            &.special-button{
                background-color: $grey;
                font-size: 21px;

                &::before{
                    background: linear-gradient(to bottom right, $grey-light, $grey, $grey-dark);
                }

                &:active{
                    font-size: 20px;
                    background-color: $grey-dark;

                    &::before{
                        background: linear-gradient(to bottom right, black, $grey);
                    }
                }
            }
        }
    }

    @media only screen and (max-width: 420px) {
        $buttons-size: 12.86vw;
        padding: $buttons-size/2;
        box-shadow: $buttons-size/4 $buttons-size/4 $buttons-size/2 rgba(black, 0.5);
        margin-top: 11.90vw;

        .display{
            font-size: 10.95vw;
            height: $buttons-size + 2.38vw;
            line-height: $buttons-size + 2.38vw;
            padding: 0 2.38vw 0 2.38vw;
            margin-bottom: $buttons-size/2;
            border-radius: 6px;
            box-shadow: 0 0 2.38vw rgba(black, 0.5) inset;
            text-shadow: 0 0 2.38vw rgba(black, 0.2);
        }

        .buttons{
            grid-gap: $buttons-size/3;
            font-size: 10.95vw;

            button{
                line-height: $buttons-size;
                height: $buttons-size;
                width: $buttons-size;


                &::before{
                    box-shadow: 2.38vw 2.38vw $buttons-size/2 $orange-dark;
                }

                &:hover{

                    &::before{
                        box-shadow: 0 0 $buttons-size/2 $orange-light;
                    }
                }

                &:active{
                    font-size: 10.48vw;
                }


                &.special-button{
                    font-size: 5vw;

                    &:active{
                        font-size: 4.76vw;
                    }
                }
            }
        }
    }
}
