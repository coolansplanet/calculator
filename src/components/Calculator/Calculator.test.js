import Calculator from './Calculator';
import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });


describe('Calculator component snapshot', () => {

    it('Matches the snapshot', () => {
        const tree = renderer.create(<Calculator/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});


describe('Calculator component click events', () => {


    let wrapper;
    const displayOutput = () => {
        return wrapper.find('.display').text();
    }
    const click = node => {
        wrapper.find(node).simulate('click');
    }

    beforeEach(() => {
        wrapper = shallow(<Calculator/>);
    })


    it('The first displayed number should be 0', () => {
        expect(
            displayOutput()
        ).toEqual('0');
    });

    it('All numbers buttons work', () => {

        for (let i = 0; i <= 9; i++) {
            click('.b'+i);
        }
        click('.b0');
        expect(
            displayOutput()
        ).toEqual('1234567890');

    });

    it('Button 0 inputs only once', () => {

        for (let i = 0; i < 100; i++) {
            click('.b0');
        }

        expect(
            displayOutput()
        ).toEqual('0');
    });


    it('Button 0 inputs only once (different test)', () => {

        click('.b0');
        click('.point');
        click('.equals');

        for (let i = 0; i < 9; i++) {
            click('.b0');
        }

        expect(
            displayOutput()
        ).toEqual('0');
    });


    it('Button 0 inputs only once (different test)', () => {

        click('.b2');
        click('.addition');

        for (let i = 0; i < 2; i++) {
            click('.b0');
        }

        expect(
            displayOutput()
        ).toEqual('0');
    });


    it('Point inputs only once', () => {

        for (let i = 0; i < 100; i++) {
            if (i == 0 || i == 50) {
                click('.b1');
            }else{
                click('.point');
            }
        }

        expect(
            displayOutput()
        ).toEqual('1.1');
    });


    it('Point clicked at first time is added after zero', () => {

        for (let i = 0; i < 100; i++) {
            click('.point');
        }
        click('.b0');

        expect(
            displayOutput()
        ).toEqual('0.0');
    });


    it('Multiplication works properly', () => {

        for (let i = 1; i <= 9; i++) {

            click('.b'+i);
            click('.multiplication');
        }

        expect(
            displayOutput()
        ).toEqual('362880');
    });


    it('Multiplication works properly', () => {

        for (let i = 1; i <= 9; i++) {
            click('.b9');
            click('.multiplication');
        }

        expect(
            displayOutput()
        ).toEqual('387420489');
    });


    it('Addition works properly', () => {

        for (let i = 1; i <= 9; i++) {
            click('.b'+i);
            click('.addition');
        }

        expect(
            displayOutput()
        ).toEqual('45');
    });


    it('Addition works properly (different addition)', () => {

        for (let i = 1; i <= 9; i++) {
            click('.b1');
            click('.addition');
            click('.b4');
            click('.addition');
            click('.b9');
            click('.addition');
        }

        expect(
            displayOutput()
        ).toEqual('126');
    });


    it('Division works properly', () => {

        click('.b7');
        click('.b2');
        click('.b9');
        click('.b8');
        click('.b5');
        click('.b1');
        click('.b4');
        click('.division');
        click('.b3');

        click('.equals');

        expect(
            displayOutput()
        ).toEqual('2432838');
    });


    it('Substraction and equals work properly', () => {

        click('.b8');
        click('.b7');

        for (let i = 0; i <= 9; i++) {

            click('.substraction');
            click('.b'+i);
        }
        click('.equals');

        expect(
            displayOutput()
        ).toEqual('42')
    });


    it('C (clear all) button works properly', () => {

        for (let i = 0; i <= 9; i++) {
            click('.addition');
            click('.b'+i);
        }
        click('.equals');
        click('.clear-all');

        expect(
            displayOutput()
        ).toEqual('0')
    });


    it('Del (delete one character) button works properly', () => {

        for (let i = 0; i <= 9; i++) {
            click('.b'+i);
        }
        click('.point');
        click('.b0');

        click('.delete');
        click('.delete');
        click('.delete');

        expect(
            displayOutput()
        ).toEqual('12345678')
    });


    it('Display don\'t display numbers containing more than 11 digits', () => {

        for (let i = 0; i <= 90; i++) {
            i === 5 ? click('.point') : click('.b8');
        }

        let output = displayOutput();
        let pointIndex = output.indexOf('.');
        output.slice(pointIndex, pointIndex+1);
        expect(
            output.length
        ).toBeLessThanOrEqual(11);
    });


    it('Display don\'t display numbers containing more than 11 digits when doing operation', () => {

        for (let i = 1; i < 100; i++) {
            click('.b9');
            click('.multiplication');
        }
        click('.equals');
        let output = displayOutput();
        let pointIndex = output.indexOf('.');
        output.slice(pointIndex, pointIndex+1);
        expect(
            output.length
        ).toBeLessThanOrEqual(11);
    });

    it('NaN displays as \'Error\'', () => {

        click('.b0');
        click('.division');
        click('.b0');
        click('.equals');
        expect(
            displayOutput()
        ).toEqual('Error');
    });


    it('Many addition inputs shouldn\'t change the display', () => {

        click('.b2');
        for (let i = 0; i < 20; i++) {
            click('.addition');
        }
        expect(
            displayOutput()
        ).toEqual('2');
    });


    it('Many multiplication inputs shouldn\'t change the display', () => {

        click('.b2');
        for (let i = 0; i < 20; i++) {
            click('.multiplication');
        }
        expect(
            displayOutput()
        ).toEqual('2');
    });


    it('Many division inputs shouldn\'t change the display', () => {

        click('.b2');
        for (let i = 0; i < 20; i++) {
            click('.division');
        }
        expect(
            displayOutput()
        ).toEqual('2');
    });


    it('Many subsraction inputs shouldn\'t change the display', () => {

        click('.b2');
        for (let i = 0; i < 20; i++) {
            click('.substraction');
        }
        expect(
            displayOutput()
        ).toEqual('2');
    });


    it('Button \'Del\' at the beninning, and twice clicks on 0 shouldn\'t show more than one zero', () => {

        click('.delete');
        click('.b0');
        click('.b0');
        expect(
            displayOutput()
        ).toEqual('0');
    });


    it('Button \'Del\' at the beninning shouldn\'t delete', () => {

        click('.delete');
        expect(
            displayOutput()
        ).toEqual('0');
    });


    it('Button \'Del\' shouldn\'t delete the last character. It must convert it to 0 instead', () => {

        click('.b1');
        click('.delete');
        expect(
            displayOutput()
        ).toEqual('0');
    });


    it('After clicking CE, if I input 0123 shouldn\'t show the number 0', () => {

        click('.clear-entry');
        click('.b0');
        click('.b1');
        click('.b2');
        click('.b3');
        expect(
            displayOutput()
        ).toEqual('123');
    });
});
