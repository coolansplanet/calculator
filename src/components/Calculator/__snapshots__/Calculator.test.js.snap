// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component snapshot Matches the snapshot 1`] = `
<div
  className="calculator"
>
  <div
    className="display"
  >
    0
  </div>
  <div
    className="buttons"
  >
    <button
      className="b7"
      onClick={[Function]}
    >
      7
    </button>
    <button
      className="b8"
      onClick={[Function]}
    >
      8
    </button>
    <button
      className="b9"
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="division"
      onClick={[Function]}
    >
      รท
    </button>
    <button
      className="special-button clear-all"
      onClick={[Function]}
    >
      C
    </button>
    <button
      className="b4"
      onClick={[Function]}
    >
      4
    </button>
    <button
      className="b5"
      onClick={[Function]}
    >
      5
    </button>
    <button
      className="b6"
      onClick={[Function]}
    >
      6
    </button>
    <button
      className="multiplication"
      onClick={[Function]}
    >
      ร
    </button>
    <button
      className="special-button clear-entry"
      onClick={[Function]}
    >
      CE
    </button>
    <button
      className="b1"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="b2"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="b3"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="substraction"
      onClick={[Function]}
    >
      -
    </button>
    <button
      className="special-button delete"
      onClick={[Function]}
    >
      DEL
    </button>
    <button
      className="b0"
      onClick={[Function]}
    >
      0
    </button>
    <button
      className="point"
      onClick={[Function]}
    >
      .
    </button>
    <button
      className="equals"
      onClick={[Function]}
    >
      =
    </button>
    <button
      className="addition"
      onClick={[Function]}
    >
      +
    </button>
    <button
      className="special-button about"
      onClick={[Function]}
    >
      A
    </button>
  </div>
</div>
`;
